File: m3_adapter/gvd/objects.py
"""objects.py — scene-graph Objects layer: cluster semantic voxels into object
instances (DBSCAN tolerance clustering per non-structure class, with bbox-proximity
post-merge to de-fragment same-class splits) and link each to its containing place
node. (Hydra Objects layer.)"""
from __future__ import annotations
from dataclasses import dataclass, field
import numpy as np

# uhumans2 structure classes (not objects): floor=3, ceiling=4, wall=19
DEFAULT_STRUCTURE = frozenset({3, 4, 19})


@dataclass
class ObjectNode:
    idx: tuple            # centroid dense voxel index
    label: int            # super_id
    label_name: str
    voxel_count: int
    bbox_min: tuple       # dense voxel
    bbox_max: tuple       # dense voxel (inclusive)
    place_id: int = -1    # nearest place node id (-1 = unassigned)
    features: dict = field(default_factory=dict)


def _bbox_gap(b1min, b1max, b2min, b2max):
    """Max axis gap (in empty voxels) between two integer-voxel bboxes.
    Overlap or touching => 0; one empty voxel between them => 1; etc.
    Formula: number of empty integer positions between the inclusive bbox edges."""
    gap = 0
    for a in range(3):
        # positive => empty voxels between the two extents; <=0 => touching/overlap
        sep = max(b2min[a] - b1max[a] - 1, b1min[a] - b2max[a] - 1, 0)
        gap = max(gap, sep)
    return gap


def _union_find_merge(clusters, merge_gap_voxels):
    """Given a list of cell arrays (one per DBSCAN cluster), merge those whose
    bounding boxes are within merge_gap_voxels of each other via union-find.
    Returns a list of merged cell arrays."""
    n = len(clusters)
    if n == 0:
        return []

    # precompute bboxes
    bmins = [c.min(axis=0) for c in clusters]
    bmaxs = [c.max(axis=0) for c in clusters]

    # union-find (path-compressed)
    parent = list(range(n))

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    def union(x, y):
        rx, ry = find(x), find(y)
        if rx != ry:
            parent[ry] = rx

    # O(n^2) pairwise; n is number of DBSCAN clusters per class — small in practice
    for i in range(n):
        for j in range(i + 1, n):
            if _bbox_gap(bmins[i], bmaxs[i], bmins[j], bmaxs[j]) <= merge_gap_voxels:
                union(i, j)

    # collect groups
    from collections import defaultdict
    groups = defaultdict(list)
    for i in range(n):
        groups[find(i)].append(i)

    merged = []
    for idxs in groups.values():
        combined = np.concatenate([clusters[i] for i in idxs], axis=0)
        merged.append(combined)
    return merged


def extract_objects(occ, sem, voxel_size, vmin, label_names=None,
                    structure_labels=DEFAULT_STRUCTURE, min_voxels=20,
                    cluster_eps_voxels: float = 3.0, min_samples: int = 5,
                    merge_gap_voxels: int = 2):
    """Cluster occupied voxels of each non-structure semantic class into object
    instances via DBSCAN tolerance clustering (bridges small geometric gaps),
    followed by a bbox-proximity merge of same-class fragments.
    Returns list[ObjectNode].

    Coordinate contract: all centroid / bbox values are in dense grid coords
    (same as the old connected-components implementation). Callers that convert
    to world metres add vmin externally — that contract is unchanged.
    """
    from m3_adapter.clustering import dbscan_labels
    from m3_adapter.gvd.features import extract_features

    label_names = label_names or {}
    structure = set(int(s) for s in structure_labels)
    objects = []
    present = [int(L) for L in np.unique(sem) if L != 0 and int(L) not in structure]

    for L in present:
        mask = occ & (sem == L)
        # argwhere over the FULL grid → coords are already full-grid indices,
        # no base-offset needed (unlike the old find_objects approach).
        cells_all = np.argwhere(mask)
        if len(cells_all) < min_voxels:
            continue

        # DBSCAN in voxel units — eps bridges gaps ≤ cluster_eps_voxels; noise (-1) dropped
        lab = dbscan_labels(cells_all, cluster_eps_voxels, min_samples)

        clusters = []
        for g in np.unique(lab):
            if g < 0:
                continue   # noise
            cells = cells_all[lab == g]
            if len(cells) < min_voxels:
                continue
            clusters.append(cells)

        if not clusters:
            continue

        # bbox-proximity merge: union-find over clusters whose bboxes are within
        # merge_gap_voxels; then build one ObjectNode per merged group
        merged_clusters = _union_find_merge(clusters, merge_gap_voxels)

        ctx = {"occ": occ, "sem": sem}
        for cells in merged_clusters:
            if len(cells) < min_voxels:
                continue
            centroid = cells.mean(axis=0).round().astype(int)
            bmin = cells.min(axis=0)
            bmax = cells.max(axis=0)
            feats = extract_features(cells, voxel_size, ctx=ctx)
            objects.append(ObjectNode(
                idx=tuple(int(x) for x in centroid),
                label=L, label_name=label_names.get(L, str(L)),
                voxel_count=len(cells),
                bbox_min=tuple(int(x) for x in bmin),
                bbox_max=tuple(int(x) for x in bmax),
                features=feats))

    return objects


def link_to_places(objects, places_graph):
    """Assign each object the nearest place node id (by centroid world-metre
    distance to place positions). Mutates objects' place_id; returns objects."""
    if not objects or not places_graph.nodes:
        return objects
    ppos = places_graph.positions_m()                  # (P,3)
    vs = places_graph.voxel_size
    vmin = np.asarray(places_graph.vmin, dtype=float)
    for o in objects:
        opos = (np.asarray(o.idx, dtype=float) + vmin) * vs
        d2 = ((ppos - opos[None, :]) ** 2).sum(axis=1)
        o.place_id = int(np.argmin(d2))
    return objects

File: semantic_map_core/src/objects.cpp
// objects.cpp — 见 objects.hpp。移植自 m3_adapter/gvd/objects.py + clustering.py。
#include "semantic_map_core/objects.hpp"

#include <algorithm>
#include <cmath>
#include <functional>
#include <unordered_map>

namespace smc {

namespace {
inline long IDX(int i, int j, int k, int ny, int nz) {
  return (static_cast<long>(i) * ny + j) * nz + k;
}
inline long ckey(int i, int j, int k) {
  const long K = 1L << 20;
  return ((long)(i + K) << 42) | ((long)(j + K) << 21) | (long)(k + K);
}
}  // namespace

std::vector<int> dbscan_labels_int3(const std::vector<std::array<int, 3>>& pts,
                                    double eps, int min_samples) {
  const int n = (int)pts.size();
  std::vector<int> labels(n, -1);
  if (n == 0) return labels;
  const double eps2 = eps * eps;
  const int R = (int)std::ceil(eps);
  std::unordered_map<long, int> h;
  h.reserve(n * 2);
  for (int p = 0; p < n; ++p) h[ckey(pts[p][0], pts[p][1], pts[p][2])] = p;

  std::vector<std::vector<int>> nbrs(n);
  std::vector<char> core(n, 0);
  for (int p = 0; p < n; ++p) {
    int i = pts[p][0], j = pts[p][1], k = pts[p][2];
    int cnt = 0;
    for (int di = -R; di <= R; ++di)
      for (int dj = -R; dj <= R; ++dj)
        for (int dk = -R; dk <= R; ++dk) {
          if ((double)(di * di + dj * dj + dk * dk) > eps2) continue;
          auto it = h.find(ckey(i + di, j + dj, k + dk));
          if (it == h.end()) continue;
          ++cnt;  // 含自身
          if (it->second != p) nbrs[p].push_back(it->second);
        }
    core[p] = cnt >= min_samples;
  }
  bool any = false;
  for (int p = 0; p < n; ++p) if (core[p]) { any = true; break; }
  if (!any) return labels;

  std::vector<int> par(n);
  for (int p = 0; p < n; ++p) par[p] = p;
  std::function<int(int)> find = [&](int x) {
    while (par[x] != x) { par[x] = par[par[x]]; x = par[x]; }
    return x;
  };
  for (int p = 0; p < n; ++p)
    if (core[p])
      for (int q : nbrs[p])
        if (core[q]) { int ra = find(p), rb = find(q); if (ra != rb) par[std::max(ra, rb)] = std::min(ra, rb); }

  std::unordered_map<int, int> r2l;
  for (int p = 0; p < n; ++p)
    if (core[p]) {
      int r = find(p);
      auto it = r2l.find(r);
      if (it == r2l.end()) { int lb = (int)r2l.size(); r2l[r] = lb; labels[p] = lb; }
      else labels[p] = it->second;
    }
  // border: 非 core 但 eps 内有 core -> 取最小 core 邻居簇号
  for (int p = 0; p < n; ++p)
    if (!core[p])
      for (int q : nbrs[p])
        if (core[q]) { int lq = labels[q]; if (labels[p] == -1 || lq < labels[p]) labels[p] = lq; }
  return labels;
}

namespace {
int bbox_gap(const int* b1min, const int* b1max, const int* b2min, const int* b2max) {
  int gap = 0;
  for (int a = 0; a < 3; ++a) {
    int sep = std::max(std::max(b2min[a] - b1max[a] - 1, b1min[a] - b2max[a] - 1), 0);
    gap = std::max(gap, sep);
  }
  return gap;
}
}  // namespace

std::vector<ObjectNode> extract_objects(const uint8_t* occ, const uint8_t* sem, int nx,
                                        int ny, int nz,
                                        const std::set<int>& structure_labels,
                                        int min_voxels, double cluster_eps, int min_samples,
                                        int merge_gap) {
  std::vector<ObjectNode> out;
  // 出现的非结构、非 0 类(升序)
  std::set<int> present;
  for (long id = 0; id < (long)nx * ny * nz; ++id)
    if (occ[id] && sem[id] != 0 && !structure_labels.count(sem[id])) present.insert(sem[id]);

  for (int L : present) {
    // cells_all = argwhere(occ & sem==L), C 序
    std::vector<std::array<int, 3>> cells;
    for (int i = 0; i < nx; ++i)
      for (int j = 0; j < ny; ++j)
        for (int k = 0; k < nz; ++k) {
          long id = IDX(i, j, k, ny, nz);
          if (occ[id] && sem[id] == L) cells.push_back({i, j, k});
        }
    if ((int)cells.size() < min_voxels) continue;

    auto lab = dbscan_labels_int3(cells, cluster_eps, min_samples);
    int maxlab = -1;
    for (int l : lab) maxlab = std::max(maxlab, l);
    // clusters: g 升序, g>=0, size>=min_voxels
    std::vector<std::vector<int>> clusters;  // 每簇存 cells 索引
    for (int g = 0; g <= maxlab; ++g) {
      std::vector<int> idxs;
      for (int p = 0; p < (int)cells.size(); ++p) if (lab[p] == g) idxs.push_back(p);
      if ((int)idxs.size() >= min_voxels) clusters.push_back(std::move(idxs));
    }
    if (clusters.empty()) continue;

    // union-find bbox 邻近合并
    int nc = (int)clusters.size();
    std::vector<std::array<int, 3>> bmin(nc), bmax(nc);
    for (int c = 0; c < nc; ++c) {
      std::array<int, 3> lo{1 << 30, 1 << 30, 1 << 30}, hi{-(1 << 30), -(1 << 30), -(1 << 30)};
      for (int pi : clusters[c])
        for (int a = 0; a < 3; ++a) { lo[a] = std::min(lo[a], cells[pi][a]); hi[a] = std::max(hi[a], cells[pi][a]); }
      bmin[c] = lo; bmax[c] = hi;
    }
    std::vector<int> par(nc);
    for (int c = 0; c < nc; ++c) par[c] = c;
    std::function<int(int)> find = [&](int x) { while (par[x] != x) { par[x] = par[par[x]]; x = par[x]; } return x; };
    for (int a = 0; a < nc; ++a)
      for (int b = a + 1; b < nc; ++b)
        if (bbox_gap(bmin[a].data(), bmax[a].data(), bmin[b].data(), bmax[b].data()) <= merge_gap) {
          int ra = find(a), rb = find(b); if (ra != rb) par[rb] = ra;  // union(a,b): parent[find(b)]=find(a)
        }
    // 分组(按 root 首次出现序), 合并 cells
    std::unordered_map<int, std::vector<int>> groups;
    std::vector<int> group_order;
    for (int c = 0; c < nc; ++c) {
      int r = find(c);
      if (!groups.count(r)) group_order.push_back(r);
      for (int pi : clusters[c]) groups[r].push_back(pi);
    }
    for (int r : group_order) {
      auto& merged = groups[r];
      if ((int)merged.size() < min_voxels) continue;
      double cx = 0, cy = 0, cz = 0;
      std::array<int, 3> lo{1 << 30, 1 << 30, 1 << 30}, hi{-(1 << 30), -(1 << 30), -(1 << 30)};
      for (int pi : merged) {
        cx += cells[pi][0]; cy += cells[pi][1]; cz += cells[pi][2];
        for (int a = 0; a < 3; ++a) { lo[a] = std::min(lo[a], cells[pi][a]); hi[a] = std::max(hi[a], cells[pi][a]); }
      }
      double m = merged.size();
      ObjectNode o;
      o.ci = (int)std::lrint(cx / m); o.cj = (int)std::lrint(cy / m); o.ck = (int)std::lrint(cz / m);
      o.label = L; o.voxel_count = (int)merged.size();
      for (int a = 0; a < 3; ++a) { o.bmin[a] = lo[a]; o.bmax[a] = hi[a]; }
      out.push_back(o);
    }
  }
  return out;
}

}  // namespace smc
